parcels/collection/collectionnodes.py

```diff
@@ -0,0 +1,1782 @@
+                    for i in range(self._pu_centers.shape[0]):
+                        ax = float(pu_ndata[i]) / float(len(np.nonzero(self._pu_indicators == i)[0]))
+                        self._pu_centers[i, :] += ax*pu_ncenters[i, :]
+                mpi_comm.Bcast(self._pu_centers, root=0)
+            else:
+                pu_indices = np.arange(start=0, stop=data_array['lon'].shape[0])
+                n_pu_data = pu_indices.shape[0]
+            if n_pu_data <= 0:
+                results = []
+            else:
+                for i in pu_indices:
+                    pdata = self._pclass(lon=data_array['lon'][i], lat=data_array['lat'][i], pid=np.iinfo(np.int64).max, ngrids=self._ngrid)
+                    if 'depth' in data_array.keys():
+                        pdata.depth = data_array['depth'][i]
+                    if 'time' in data_array.keys():
+                        pdata.time = data_array['time'][i]
+                    if 'dt' in data_array.keys():
+                        pdata.dt = data_array['dt'][i]
+                    for key in self._kwarg_keys:
+                        if key in data_array.keys():
+                            setattr(pdata, key, data_array[key][i])
+                    results.append(self.add_single(pdata, pu_checked=True))
+        self._ncount = len(self._data)
+        return results
+
+    def add_single(self, particle_obj, pu_checked=False):
+        """
+        Adding a single Particle to the collection - either as a 'Particle' object in parcels itself, or
+        via its ParticleAccessor.
+        :returns index of added particle
+        """
+        # ==== first approach - still need to incorporate the MPI re-centering ==== #
+        super().add_single(particle_obj)
+        assert isinstance(particle_obj, ScipyParticle)
+        # Comment: by current workflow, pset modification is only done on the front node, thus
+        # the distance determination and assigment is also done on the front node
+        _add_to_pu = True
+        if MPI and MPI.COMM_WORLD.Get_size() > 1 and not pu_checked:
+            if self._pu_centers is not None and isinstance(self._pu_centers, np.ndarray):
+                mpi_comm = MPI.COMM_WORLD
+                mpi_rank = mpi_comm.Get_rank()
+                mpi_size = mpi_comm.Get_size()
+                min_dist = np.finfo(self._lonlatdepth_dtype).max
+                min_pu = 0
+                spdata = None
+                if mpi_size > 1 and mpi_rank == 0:
+                    ppos = particle_obj
+                    if isinstance(particle_obj, self._nclass):
+                        ppos = particle_obj.data
+                    spdata = np.array([ppos.lat, ppos.lon], dtype=self._lonlatdepth_dtype)
+                    n_clusters = self._pu_centers.shape[0]
+                    for i in range(n_clusters):
+                        diff = self._pu_centers[i, :] - spdata
+                        dist = np.dot(diff, diff)
```

Comment:
I would recommend to separate all this distance calculation into a separate tool (class?) for easy swapping between e.g. spherical and flat meshes and also to try to accommodate for periodic boundary conditions

Reply:
Now a separate issue #1117